src/components/focus-manager/useFocusManager.tsx

```diff
@@ -4,52 +4,91 @@
  *
  * (c) 2021 Feedzai, Rights Reserved.
  */
+import { MutableRefObject, RefObject, useMemo, useRef } from "react";
+import { createFocusTrap } from "focus-trap";
+import { FocusManagerProps, UseFocusManagerAsProps, UseFocusManagerWithRef } from "./types";
+import { getFallbackFocus, getInitialFocus, getReturnFocusProps } from "./helpers";
 
 /**
- * useFocusManager.tsx
- *
- * @author João Dias <joao.dias@feedzai.com>
- * @since 1.0.0
- */
-import { useContext } from "react";
-import { IFocusManager } from "./index";
-import { FocusContext } from "./context";
-
-/**
- * Returns a FocusManager interface for the parent FocusScope.
+ * Returns an FocusManager interface for the parent FocusManager.
  * A FocusManager can be used to programmatically move focus within
- * a FocusScope, e.g. in response to user events like keyboard navigation.
+ * a FocusManager, e.g. in response to user events like keyboard navigation.
  *
  * @example
- * function Toolbar(props) {
- * 	return (
- *  	<div role="toolbar"><FocusScope>{props.children}</FocusScope></div>
- *  );
- * }
- *
- * function ToolbarButton(props) {
- * 	const focusManager = useFocusManager();
- * 	const onKeyDown = (event) => {
- * 	switch (event.key) {
- * 		case 'ArrowRight':
- * 			focusManager.focusNext({ wrap: true });
- * 			break;
- * 		case 'ArrowLeft':
- * 			focusManager.focusPrevious({ wrap: true });
- * 			break;
- * 	}
- * };
+ * ```tsx
+ * import { useFocusManager } from "@feedzai/react-a11y-tools";
+ * ...
+ * const elementRef = useRef(null);
  *
- * return <button onKeyDown={onKeyDown}>{props.children}</button>;}
+ * useFocusManager(elementRef)
  *
- * <Toolbar>
- * 	<ToolbarButton>Cut</ToolbarButton>
- * 	<ToolbarButton>Copy</ToolbarButton>
- * 	<ToolbarButton>Paste</ToolbarButton>
- * </Toolbar>
+ * return (
+ *  <div ref={elementRef}>
+ *   <input data-testid="fdz-js-input-1" />
+ *   <input data-testid="fdz-js-input-2" />
+ *   <input data-testid="fdz-js-input-3" />
+ *  </div>
+ * )
+ * ```
  *
- * @returns {IFocusManager}
  */
-export function useFocusManager(): IFocusManager | null {
-	return useContext(FocusContext);
+export function useFocusManager<GenericElement extends HTMLElement | SVGElement>(
+	props: Omit<FocusManagerProps, "children">,
+	ref: RefObject<GenericElement> | MutableRefObject<GenericElement>,
+): UseFocusManagerWithRef;
+export function useFocusManager(
+	props: Omit<FocusManagerProps, "children">,
+	ref?: never,
+): UseFocusManagerAsProps;
+export function useFocusManager<GenericElement extends HTMLElement | SVGElement>(
```

Comment:
Was this a mistake? Exporting 3 functions with the name `useFocusManager` on the same file.

Reply:
Nope, it is a typescript feature [called function overloading](https://blog.logrocket.com/implementing-function-overloading-typescript/).
Basically, you can define multiple function signatures so that, depending on the input, the expected output is different. 
For example, if you pass only the props, the hook behaves as if inside a component, otherwise it will behave like a hook that provides the `FocusTrap` functionality.